File: oferta_clase.php
<?php
ob_start();
require 'menu.php';
require 'db.php';
require_once __DIR__ . '/lib/csrf.php';

require_once __DIR__ . '/lib/security_headers.php';

if (!isset($_SESSION['usuarioId'])) { header('Location: login.php'); exit; }
$uid = (int)$_SESSION['usuarioId'];

$error   = '';
$success = '';

// Load subjects and teacher's currency
$materias = dbAll("SELECT materiaId, nombre FROM materias ORDER BY orden ASC");
$teacher  = dbOne(
    "SELECT u.rol, pa.nombre AS pais, pa.simbolo, pa.codigo_moneda, pa.tasa_usd
     FROM usuarios u LEFT JOIN paises pa ON pa.paisId = u.pais_id
     WHERE u.usuarioId = :id",
    ['id' => $uid]
);

// â”€â”€ SAVE FILTER / CLASS OFFER â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    csrf_require();
    $precio_min  = max(0, (float)($_POST['precio_min']  ?? 0));
    $precio_max  = max(0, (float)($_POST['precio_max']  ?? 0));
    $alumnos_min = max(1, (int)($_POST['alumnos_min']   ?? 1));
    $alumnos_max = max(1, (int)($_POST['alumnos_max']   ?? 1));
    $materiaId   = (int)($_POST['materiaId'] ?? 0) ?: null;
    $solo_yo     = isset($_POST['solo_yo']) ? 1 : 0;
    $moneda      = $teacher['codigo_moneda'] ?? 'USD';

    // Use midpoint as base price
    $precio_base = $precio_min > 0 ? $precio_min : max(1, $precio_max);

    if ($precio_max > 0 && $precio_max < $precio_min) {
        $error = t('oferta.error_price');
    } elseif ($alumnos_max < $alumnos_min) {
        $error = t('oferta.error_students');
    } else {
        $claseId = dbExec(
            "INSERT INTO clases_programadas
                 (instructorId, materiaId, precio_min, precio_max, precio_base,
                  codigo_moneda, alumnos_min, alumnos_max, solo_yo, activa)
             VALUES (:i, :m, :pmin, :pmax, :pbase, :mon, :amin, :amax, :solo, 1)",
            [
                'i'     => $uid,
                'm'     => $materiaId,
                'pmin'  => $precio_min,
                'pmax'  => $precio_max,
                'pbase' => $precio_base,
                'mon'   => $moneda,
                'amin'  => $alumnos_min,
                'amax'  => $alumnos_max,
                'solo'  => $solo_yo,
            ]
        );
        if ($claseId) {
            $success = t('oferta.success');
        } else {
            $error = t('oferta.error_db');
        }
    }
}
?>

  <div class="container mt-10">
    <div class="row justify-content-center">
      <div class="col-sm-10 col-md-7 col-lg-6">

        <h3 class="text-white mb-1"><?= t('oferta.title') ?></h3>
        <p class="text-secondary mb-4"><?= t('oferta.subtitle') ?></p>

        <?php if ($error): ?>
          <div class="alert alert-danger"><?= htmlspecialchars($error) ?></div>
        <?php endif; ?>
        <?php if ($success): ?>
          <div class="alert alert-success"><?= htmlspecialchars($success) ?></div>
        <?php endif; ?>

        <?php if ($teacher): ?>
          <div class="alert alert-dark border border-secondary mb-4 small">
            💰 <?= t('oferta.your_currency') ?>
            <strong class="text-white"><?= htmlspecialchars($teacher['simbolo'] . ' ' . $teacher['codigo_moneda']) ?></strong>
            · <?= t('oferta.country') ?> <strong class="text-white"><?= htmlspecialchars($teacher['pais'] ?? '”') ?></strong>
          </div>
        <?php endif; ?>

        <form method="POST" action="oferta_clase.php" class="card bg-dark border-secondary p-4">
<?= csrf_field() ?>

          <!-- Subject -->
          <div class="mb-3">
            <label class="form-label text-secondary"><?= t('oferta.subject') ?> <span class="text-secondary"><?= t('oferta.optional') ?></span></label>
            <select name="materiaId" class="form-select bg-dark text-white border-secondary">
              <option value=""><?= t('oferta.any_subject') ?></option>
              <?php foreach ($materias as $m): ?>
                <option value="<?= $m['materiaid'] ?>"
                        <?= (int)($_POST['materiaId'] ?? 0) === (int)$m['materiaid'] ? 'selected' : '' ?>>
                  <?= htmlspecialchars($m['nombre']) ?>
                </option>
              <?php endforeach; ?>
            </select>
          </div>

          <!-- Price range -->
          <div class="row g-3 mb-3">
            <div class="col-6">
              <label class="form-label text-secondary"><?= t('oferta.min_price') ?> (<?= htmlspecialchars($teacher['codigo_moneda'] ?? 'USD') ?>)</label>
              <input type="number" name="precio_min" class="form-control bg-dark text-white border-secondary"
                     placeholder="0" min="0" step="0.01"
                     value="<?= htmlspecialchars($_POST['precio_min'] ?? '') ?>">
            </div>
            <div class="col-6">
              <label class="form-label text-secondary"><?= t('oferta.max_price') ?> (<?= htmlspecialchars($teacher['codigo_moneda'] ?? 'USD') ?>)</label>
              <input type="number" name="precio_max" class="form-control bg-dark text-white border-secondary"
                     placeholder="0" min="0" step="0.01"
                     value="<?= htmlspecialchars($_POST['precio_max'] ?? '') ?>">
            </div>
          </div>

          <!-- Student count -->
          <div class="row g-3 mb-3">
            <div class="col-6">
              <label class="form-label text-secondary"><?= t('oferta.min_students') ?></label>
              <input type="number" name="alumnos_min" class="form-control bg-dark text-white border-secondary"
                     placeholder="1" min="1" step="1"
                     value="<?= htmlspecialchars($_POST['alumnos_min'] ?? '1') ?>">
            </div>
            <div class="col-6">
              <label class="form-label text-secondary"><?= t('oferta.max_students') ?></label>
              <input type="number" name="alumnos_max" class="form-control bg-dark text-white border-secondary"
                     placeholder="10" min="1" step="1"
                     value="<?= htmlspecialchars($_POST['alumnos_max'] ?? '10') ?>">
            </div>
          </div>

          <!-- Solo -->
          <div class="mb-4 form-check">
            <input class="form-check-input" type="checkbox" name="solo_yo" id="solo_yo"
                   <?= isset($_POST['solo_yo']) ? 'checked' : '' ?>>
            <label class="form-check-label text-secondary" for="solo_yo">
              <?= t('oferta.one_on_one') ?>
            </label>
          </div>

          <div class="d-flex gap-2">
            <button type="submit" class="btn btn-secondary flex-grow-1"><?= t('oferta.save') ?></button>
            <a href="crear_clase.php" class="btn btn-dark border-secondary"><?= t('oferta.add_title') ?></a>
          </div>
        </form>

      </div>
    </div>
  </div>

  <footer class="mastfoot mt-auto mt-5">
    <div class="inner float-end">
      <p><?= t('general.footer_text', ['bootstrap' => '<a href="https://getbootstrap.com/">Bootstrap</a>', 'author' => '<a href="https://www.facebook.com/rodrigo.alejandro.1848816?locale=es_LA">@RodrigoConejeros</a>']) ?></p>
    </div>
  </footer>

  <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-YvpcrYf0tY3lHB60NNkmXc5s9fDVZLESaAA55NDzOxhy9GkcIdslK1eN7N6jIeHz" crossorigin="anonymous"></script>
  <script src="https://code.jquery.com/jquery-3.7.1.min.js" integrity="sha256-/JqT3SQfawRcv/BIHPThkBvs0OEvtFFmqPF/lYI/Cxo=" crossorigin="anonymous"></script>

  <script type="text/javascript" src="./script.js"></script>
</body>
</html>

File: go/internal/web/oferta_clase.go
package web

import (
	"net/http"

	"classexpress/internal/i18n"
	"classexpress/internal/store"
)

type materiaOption struct {
	ID       int64
	Nombre   string
	Selected bool
}

// HandleOfertaClase ports oferta_clase.php (create a class offer).
func (p *Pages) HandleOfertaClase(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	s := SessionFrom(ctx)
	if s == nil {
		serverError(w, errNoSession)
		return
	}
	if !p.GuardPage(w, r, s) {
		return
	}
	lang := p.ResolveLang(s, r)
	page := CurrentPage(r)
	nav, stop := p.MenuData(w, r, s, page, lang)
	if stop {
		return
	}
	uid := UID(s)

	errorMsg, successMsg := "", ""

	if r.Method == http.MethodPost {
		if !p.RequireCSRFOnPost(w, r, s) {
			return
		}
		precioMin := maxFloat(0, r.PostFormValue("precio_min"))
		precioMax := maxFloat(0, r.PostFormValue("precio_max"))
		alumnosMin := maxInt(1, r.PostFormValue("alumnos_min"))
		alumnosMax := maxInt(1, r.PostFormValue("alumnos_max"))
		materiaId := store.Int(r.PostFormValue("materiaId"))
		soloYo := 0
		if r.PostFormValue("solo_yo") != "" {
			soloYo = 1
		}

		teacher, err := p.DB.QueryOne(ctx,
			`SELECT u.rol, pa.codigo_moneda FROM usuarios u
			 LEFT JOIN paises pa ON pa.paisId = u.pais_id WHERE u.usuarioId = ?`, uid)
		moneda := "USD"
		if err == nil && teacher != nil {
			if c := store.Str(teacher["codigo_moneda"]); c != "" {
				moneda = c
			}
		}

		precioBase := precioMin
		if precioBase <= 0 {
			precioBase = maxFloat(1, r.PostFormValue("precio_max"))
		}

		if precioMax > 0 && precioMax < precioMin {
			errorMsg = i18n.T(lang, "oferta.error_price", nil)
		} else if alumnosMax < alumnosMin {
			errorMsg = i18n.T(lang, "oferta.error_students", nil)
		} else {
			_, err := p.DB.Exec(ctx,
				`INSERT INTO clases_programadas
				     (instructorId, materiaId, precio_min, precio_max, precio_base,
				      codigo_moneda, alumnos_min, alumnos_max, solo_yo, activa)
				 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 1)`,
				uid, nullableID(materiaId), precioMin, precioMax, precioBase,
				moneda, alumnosMin, alumnosMax, soloYo)
			if err == nil {
				successMsg = i18n.T(lang, "oferta.success", nil)
			} else {
				errorMsg = i18n.T(lang, "oferta.error_db", nil)
			}
		}
	}

	materias, err := p.DB.QueryAll(ctx, "SELECT materiaId, nombre FROM materias ORDER BY orden ASC")
	if err != nil {
		serverError(w, err)
		return
	}
	teacher, err := p.DB.QueryOne(ctx,
		`SELECT u.rol, pa.nombre AS pais, pa.simbolo, pa.codigo_moneda, pa.tasa_usd
		 FROM usuarios u LEFT JOIN paises pa ON pa.paisId = u.pais_id
		 WHERE u.usuarioId = ?`, uid)
	if err != nil {
		serverError(w, err)
		return
	}

	var opts []materiaOption
	selected := store.Int(r.PostFormValue("materiaId"))
	for _, m := range materias {
		id := store.Int(m["materiaId"])
		opts = append(opts, materiaOption{ID: id, Nombre: store.Str(m["nombre"]), Selected: selected == id})
	}

	moneda := "USD"
	simbolo := "$"
	pais := ""
	if teacher != nil {
		moneda = store.Str(teacher["codigo_moneda"])
		if moneda == "" {
			moneda = "USD"
		}
		simbolo = store.Str(teacher["simbolo"])
		pais = store.Str(teacher["pais"])
	}

	data := map[string]any{
		"Lang":          lang,
		"NavData":       nav,
		"Error":         errorMsg,
		"Success":       successMsg,
		"Materias":      opts,
		"Pais":          pais,
		"Simbolo":       simbolo,
		"Moneda":        moneda,
		"HasTeacher":    teacher != nil,
		"PrecioMin":     r.PostFormValue("precio_min"),
		"PrecioMax":     r.PostFormValue("precio_max"),
		"AlumnosMin":    r.PostFormValue("alumnos_min"),
		"AlumnosMax":    r.PostFormValue("alumnos_max"),
		"SoloYo":        r.PostFormValue("solo_yo") != "",
	}
	if err := p.Templates.RenderAuthed(w, "oferta_clase", p, s, lang, data); err != nil {
		serverError(w, err)
	}
}

func nullableID(v int64) any {
	if v <= 0 {
		return nil
	}
	return v
}

func maxFloat(def float64, v string) float64 {
	f := store.Float(v)
	if f < def {
		return def
	}
	return f
}

func maxInt(def int64, v string) int64 {
	i := store.Int(v)
	if i < def {
		return def
	}
	return i
}
